invokeai/app/services/shared/graph.py

```diff
@@ -802,6 +809,36 @@ class GraphExecutionState(BaseModel):
         description="The map of original graph nodes to prepared nodes",
         default_factory=dict,
     )
+    # Ready queues grouped by node class name (internal only)
+    _ready_queues: dict[str, Deque[str]] = PrivateAttr(default_factory=dict)
+    # Current class being drained; stays until its queue empties
+    _active_class: Optional[str] = PrivateAttr(default=None)
+    # Optional priority; others follow in name order
+    ready_order: list[str] = Field(default_factory=list)
+    indegree: dict[str, int] = Field(default_factory=dict, description="Remaining unmet input count for exec nodes")
+
+    def _type_key(self, node_obj: BaseInvocation) -> str:
+        return node_obj.__class__.__name__
+
+    def _queue_for(self, cls_name: str) -> Deque[str]:
+        q = self._ready_queues.get(cls_name)
+        if q is None:
+            q = deque()
+            self._ready_queues[cls_name] = q
+        return q
+
+    def set_ready_order(self, order: Iterable[Type[BaseInvocation] | str]) -> None:
+        names: list[str] = []
+        for x in order:
+            names.append(x.__name__ if hasattr(x, "__name__") else str(x))
+        self.ready_order = names
+
+    def _enqueue_if_ready(self, nid: str) -> None:
+        """Push nid to its class queue if unmet inputs == 0."""
+        if self.indegree.get(nid, 1) != 0 or nid in self.executed:
+            return
+        node_obj = self.execution_graph.nodes[nid]
```

Comment:
And here `nid` is assumed to exist in `self.execution_graph_nodes[]`

Reply:
Fixing this. It should be an invariant.